asyncpg/protocol/coreproto.pyx

```diff
@@ -810,34 +810,29 @@ cdef class CoreProtocol:
         self.transport.writelines(buffers)
 
     cdef _execute_many_rollback(self, object error):
-        cdef:
-            WriteBuffer packet
-            WriteBuffer buf
+        cdef WriteBuffer buf
 
         self._ensure_connected()
         if self.state != PROTOCOL_CANCELLED:
             self._set_state(PROTOCOL_BIND_EXECUTE_MANY_ROLLBACK)
-        packet = WriteBuffer.new()
-
-        # We have no idea if we are in an explicit transaction or not,
-        # therefore we raise an exception in the database to mark current
-        # transaction as failed anyway to prevent partial commit
-        buf = self._build_parse_message('',
-                                        'DO language plpgsql $$ BEGIN '
-                                        'RAISE transaction_rollback; '
-                                        'END$$;')
-        packet.write_buffer(buf)
-        buf = self._build_bind_message('', '', WriteBuffer.new())
-        packet.write_buffer(buf)
-        buf = self._build_execute_message('', 0)
-        packet.write_buffer(buf)
-        packet.write_bytes(SYNC_MESSAGE)
-
-        self.transport.write(memoryview(packet))
 
         self.result_type = RESULT_FAILED
         self.result = error
 
+        # We shall rollback in an implicit transaction to prevent partial
+        # commit, while do nothing in an explicit transaction and leaving the
+        # error to the user
+        if self.xact_status == PQTRANS_IDLE:
```

Comment:
I'd move `is_in_transaction()` to `coreproto.pyx` (where it belongs) and use that instead here.